File: flutter_client/lib/models/geo_feature.dart
// ABOUTME: Geographic feature model for markers and spatial data.
// ABOUTME: Extracted from tool results and displayed on the map.

/// Geographic feature extracted from tool results.
class GeoFeature {
  final String id;
  final String type;
  final double lat;
  final double lon;
  final String? label;

  GeoFeature({
    required this.id,
    required this.type,
    required this.lat,
    required this.lon,
    this.label,
  });

  factory GeoFeature.fromJson(Map<String, dynamic> json) {
    return GeoFeature(
      id: json['id'] as String,
      type: json['type'] as String? ?? 'marker',
      lat: (json['lat'] as num?)?.toDouble() ?? 0.0,
      lon: (json['lon'] as num?)?.toDouble() ?? 0.0,
      label: json['label'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type,
      'lat': lat,
      'lon': lon,
      'label': label,
    };
  }
}

File: agent_sdk_server/src/models/geo-feature.ts
// ABOUTME: Database model for geographic features
// ABOUTME: Stores spatial data using PostGIS geometry types

import { getPool } from '../config/database.js';

export interface GeoFeature {
  id: string;
  type: string;
  lat: number;
  lon: number;
  label: string;
}

export interface StoredGeoFeature {
  id: string;
  message_id: string;
  feature_type: string;
  geometry: any;
  properties: Record<string, any>;
  created_at: Date;
}

export async function createGeoFeature(
  messageId: string,
  feature: GeoFeature
): Promise<StoredGeoFeature> {
  const pool = getPool();

  // Convert lat/lon to PostGIS Point geometry
  const result = await pool.query(
    `INSERT INTO geo_features (id, message_id, feature_type, geometry, properties)
     VALUES ($1, $2, $3, ST_SetSRID(ST_MakePoint($4, $5), 4326), $6)
     RETURNING id, message_id, feature_type,
               ST_AsGeoJSON(geometry)::json as geometry,
               properties, created_at`,
    [feature.id, messageId, feature.type, feature.lon, feature.lat, JSON.stringify({ label: feature.label })]
  );

  return result.rows[0];
}

export async function getMessageGeoFeatures(
  messageId: string
): Promise<StoredGeoFeature[]> {
  const pool = getPool();
  const result = await pool.query(
    `SELECT id, message_id, feature_type,
            ST_AsGeoJSON(geometry)::json as geometry,
            properties, created_at
     FROM geo_features
     WHERE message_id = $1`,
    [messageId]
  );
  return result.rows;
}

export async function getConversationGeoFeatures(
  conversationId: string
): Promise<StoredGeoFeature[]> {
  const pool = getPool();
  const result = await pool.query(
    `SELECT gf.id, gf.message_id, gf.feature_type,
            ST_AsGeoJSON(gf.geometry)::json as geometry,
            gf.properties, gf.created_at
     FROM geo_features gf
     JOIN messages m ON gf.message_id = m.id
     WHERE m.conversation_id = $1
     ORDER BY gf.created_at`,
    [conversationId]
  );
  return result.rows;
}
